### PR TITLE
Changes to support the CRC32 in PowerPC.

### PR DESCRIPTION
**Pull request by MeenaRenganathan22**
Changes to support the CRC32 generation in PowerPC to address the WeakHash32 collision.
Updated the functional test 01016_simhash_minhash to support the CRC32 values based on the platform.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
CRC32 changes to address the WeakHash collision issue in PowerPC.

### Documentation entry for user-facing changes

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: contrib/crc32-vpmsum-cmake/vec_crc32.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+unsigned int crc32_vpmsum(unsigned int crc, const unsigned char *p, unsigned long len);
+
+static inline uint32_t crc32_ppc(uint64_t crc, unsigned char const *buffer, size_t len)
+{
+	unsigned char *emptybuffer;
+    if (!buffer) {
```
Comment (rschu1ze) [inline contrib/crc32-vpmsum-cmake/vec_crc32.h:14]: Sure that we should allow the caller to pass an empty buffer? Doing an on-the-fly allocation for each CRC calculation sounds quite inefficient (vs. re-using the same buffer). Could you please check if the CRC implementations on other platforms have the same logic?

File: src/Common/HashTable/Hash.h [hunk 2]
```diff
@@ -89,6 +93,8 @@ inline DB::UInt64 intHashCRC32(DB::UInt64 x)
     return __crc32cd(-1U, x);
 #elif defined(__s390x__) && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
     return s390x_crc32(-1U, x)
+#elif (defined(__PPC64__) || defined(__powerpc64__)) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Comment (rschu1ze) [inline src/Common/HashTable/Hash.h:96]: Cosmetic: Let's switch l. 94-95 and l. 96-97. (PPC is probably more important going forward than s390x). Same applies to the other files you changed.

**Review by rschu1ze — COMMENTED** [review]
I'll look at the tests later.

File: contrib/crc32-vpmsum-cmake/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+if (NOT ARCH_PPC64LE)
+    message(STATUS "crc32-vpmsum library is only supported on ppc64le")
```
[the remark this answers is not in the document]
Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/CMakeLists.txt:2]: Ok, will get it changed.

File: contrib/crc32-vpmsum-cmake/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# To Generate crc32_constants.h 
+
+- Run make file in `../crc32-vpmsum` diretory using folling options and CRC polynomial. These options should use the same polynomial and order used by intel intrinisic functions
```
[the remark this answers is not in the document]
Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/README.md:3]: Agree

File: contrib/crc32-vpmsum-cmake/vec_crc32.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#ifndef VEC_CRC32
+#define VEC_CRC32
+
```
[the remark this answers is not in the document]
Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/vec_crc32.h:3]: Ok.

File: contrib/crc32-vpmsum-cmake/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/CMakeLists.txt:2]: Changing the message category from STATUS to RECONFIGURE_MESSAGE_LEVEL will break the cross compilation.

File: contrib/crc32-vpmsum-cmake/vec_crc32.h (hunk 1 above)
Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/vec_crc32.h:14]: The above code has been added to handle the error handling. In the other CRC implementations we don't see the same logic.

Comment (rschu1ze) [reply contrib/crc32-vpmsum-cmake/vec_crc32.h:14]: Are you saying that the on-the-fly allocation block can go away? If not, could you please give an example where such code is actually needed?

(Asking because for platforms x86 and ARM, intrinsic are called (`_mm_crc32_u64`/`__crc32cd`) which don't have such logic. Perhaps you could add an `assert` that buffer is set.)

Comment (MeenaRenganathan22) [reply contrib/crc32-vpmsum-cmake/vec_crc32.h:14]: Yes, we will replace the error handle with assert.

File: tests/queries/0_stateless/01016_simhash_minhash.python [hunk 6]
```diff
@@ -0,0 +1,394 @@
+#!/usr/bin/env python3                                                                                                                                                                                              
+import os
+import socket
+import sys
+from scipy import stats
+import pandas as pd
```
Comment (rschu1ze) [inline tests/queries/0_stateless/01016_simhash_minhash.python:6]: More importantly, what I don't understand is why the expected results on PPC are different than on x86. Perhaps I missed something but was the whole point of the PR not to obtain CRC values that are equal across all platforms, given the same inputs?

If that is truly by design and cannot be changed, then there are better ways to test platform-specific behavior. Changing the SQL test into a Python test blows it up in a quite unaesthetic way. If you decide to go this route, then at least l. 28 - 183 should be factorized into a separate Python module, as they are copypasta from 02458_insert_select_progress_tcp and 02010_lc_native.python and at some point the dissemination of this boilerplate code needs to stop.

But there is no reason to mess with the native communication protocol at all for the purpose of testing CRCs. You could copy the original SQL test into two tests
- 01016_simhash_minhash.sql and
- 01016_simhash_minhash_ppc.sql

(and each SQL test will have a correspondingly named ".reference" file)

The first test is for all platforms except PPC, whereas the second test is just for PPC. Since they are copypasted, they will run the same queries (and ideally, they contain a comment at the top noting this fact and that both tests should be kept in-sync).

We can use the "Tags:" feature of SQL tests to restrict the platforms on which the 1st and 2nd tests will run. Search for this term in "tests/queries/0_stateless". Right now, tag "no-cpu-aarch64" is used in some places (to exclude ARM systems in CI). Tests in CI are executed by script "tests/clickhouse-test" and the code in there which evaluates the "no-" flag is
```py
elif tags:                                                        
│   for build_flag in args.build_flags:                           
│   │   if "no-" + build_flag in tags:                            
│   │   │   return FailureReason.BUILD                            
│   for tag in tags:                                              
│   │   tag = tag.replace("-", "_")                               
│   │   if tag.startswith("use_") and tag not in args.build_flags:
│   │   │   return FailureReason.BUILD                            
```

There are no PPC machines in ClickHouse's CI landscape (*). Basically, we want to
- exclude the 1st test from running on PPC (which is due to (*) an academic concern but nice to have) and
- exclude the 2nd test from running on x86 and ARM.

So the question is which tags must be supplied to both tests. For that, you could check clickhouse-test, specifically its method `collect_build_flags() and what is returned by ClickHouse's system table on PPC. I would like to help with that but I don't have easy access to a PPC machine.

Comment (MeenaRenganathan22) [reply tests/queries/0_stateless/01016_simhash_minhash.python:6]: > More importantly, what I don't understand is why the expected results on PPC are different than on x86. Perhaps I missed something but was the whole point of the PR not to obtain CRC values that are equal across all platforms, given the same inputs?

Power's CRC32 vpmsum uses the  CRC table same as Intel and also uses the same polynomial. And the CRC table is the software fallback version and if the hardware has the crc32c, use the crc32-vpmsum for crc32 generation. Due to the difference in hardware, we see the different hash values.

Below is the transcript in telegram:
Meena Renganathan, [Dec 9, 2022 at 10:41:07 AM]:
It is CRC32C function for Power. The CRC table follows the same approach as Intel and also uses the same polynomial

Alexey Milovidov, [Dec 9, 2022 at 10:46:38 AM]:
Then why the results are different?

Meena Renganathan, [Dec 9, 2022 at 11:27:30 AM]:
The above table is the software fallback version and if the hardware has the crc32c, use the crc32-vpmsum for crc32 generation. May be due to the difference in hardware, we see the different hash values

Alexey Milovidov, [Dec 9, 2022 at 12:39:53 PM]:
Ok. ngram difference is an approximate function, let's consider the results on different platforms ok so far, but we can change it later.

Comment (MeenaRenganathan22) [reply tests/queries/0_stateless/01016_simhash_minhash.python:6]: > We can use the "Tags:" feature of SQL tests to restrict the platforms on which the 1st and 2nd tests will run. Search for this term in "tests/queries/0_stateless". Right now, tag "no-cpu-aarch64" is used in some places (to exclude ARM systems in CI). Tests in CI are executed by script "tests/clickhouse-test" and the code in there which evaluates the "no-" flag is

Will start working on the above test approach.

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
Thanks! I'll merge as soon as the CI tests finished.

**Comment by rschu1ze** [comment]
Just noting, since CRCs are also used to checksum the database persistence, we lose the option to easily exchange data between non-PPC and PPC architectures. We discussed this internally yesterday and think this is okay, i.e. more of an edge case.

As desired, 01016_simhash_minhash ran in CI whereas 01016_simhash_minhash_ppc was skipped (--> [here](https://s3.amazonaws.com/clickhouse-test-reports/45144/983cac0a64aa3761240e9a537f8007fb63554594/fast_test.html)).

**Failing Tests**

- test_azure_blob_storage_zero_copy_replication/test.py::test_zero_copy_replication: #44862
- test_cluster_copier/test.py::test_copy_month_to_week_partition_with_recovering: #30399
